### PR TITLE
Rewrite controlnet to new model manager

### PR DESCRIPTION
**Pull request by StAlKeR7779**

### CODE REVIEW

**Review by lstein — APPROVED** [review]
LGTM

**Comment by psychedelicious** [comment]
Added the controlnet field to node editor, and your draft does seem to work fine. 

One thing we want to do is associate controlnet models to processors - like canny model defaults to canny preprocessor. I think this association should probably be in the frontend, rather than introducing a new concept to the backend. 

Any thoughts or concerns there?

Once we decide how to proceed with that, I can get the linear UI controlnet updated.

**Comment by lstein** [comment]
> Added the controlnet field to node editor, and your draft does seem to work fine.
> 
> One thing we want to do is associate controlnet models to processors - like canny model defaults to canny preprocessor. I think this association should probably be in the frontend, rather than introducing a new concept to the backend.
> 
> Any thoughts or concerns there?
> 
> Once we decide how to proceed with that, I can get the linear UI controlnet updated.

Any resolution on this decision point?

**Comment by maryhipp** [comment]
Just pushed up some changes for linear workflow - @psychedelicious I actually couldn't figure out how your node implementation was working (I couldn't get network calls to trigger and it seems to use some phantom list of constants that I can't find in the code). My changes:

- removed hard-coded list of control net models based on discord conversation
- updated controlnet model component to pull list from backend
- updated controlnet model component to disable models that have incompatible base models with selected main model
- updated processor logic to make an educated guess of processor based on substrings (I am sure this can be improved)
- moved processor + autoconfigure toggle to be further up in the UI since we are more likely to not get default correct
- fixed logic to clear submodels across the board when main base model changes (still would like to make the modal to confirm before doing this)

File: invokeai/app/invocations/controlnet_image_processors.py [hunk 1]
```diff
@@ -182,7 +190,11 @@ def invoke(self, context: InvocationContext) -> ControlOutput:
         return ControlOutput(
             control=ControlField(
                 image=self.image,
-                control_model=self.control_model,
+                #control_model=self.control_model,
+                control_model=ControlNetModelField(
+                    model_name="canny",
```
Comment (maryhipp) [inline invokeai/app/invocations/controlnet_image_processors.py:195]: @StAlKeR7779 is "canny" hard-coded here for testing purposes? what do we need to get this working with the passed-in model?

**Comment by psychedelicious** [comment]
[feat(ui): fix controlNet models](https://github.com/invoke-ai/InvokeAI/pull/3665/commits/69ac89c26235f3e283139206b05728c01401d6df)

- update controlnet state to use object format for model
- update model-parsing helper functions to log errors
- update nodes components, types and state
- remove controlnets from state when models are loaded and the controlnet's model is not available

File: invokeai/app/invocations/controlnet_image_processors.py (hunk 1 above)
Comment (psychedelicious) [reply invokeai/app/invocations/controlnet_image_processors.py]: must have been for testing, fixed

**Comment by blessedcoolant** [comment]
This PR is good to go. Maybe just changing the assertion of the control_weight to take values up to 2 instead of 1 and updating the linear UI slider to be 0, 1, 2 with 1 as default. But I'll let @psychedelicious @GreggHelt2 make the final call on that one.

Everything else seems to be working fine. Great job guys.

**Review by blessedcoolant — APPROVED** [review]

**Comment by psychedelicious** [comment]
- Updated layout
- No more `single line` layout. It's mini or full
- Did some minor styling tweaks, like fixing spacings and stuff
- Better error handling in reducers and overall logic
- fix disabled state for mantine select components
- Range is `0 - 1 - 2`
- It was not clear when a specific controlnet is enabled, because its settings are always enabled. Disable a controlnet's controls if the controlnet itself is disabled. It's still not totally clear, because the overall controlnet's enabled toggle is easy to miss, but at least this is a step in the right direction.
- Promote controlnet to be just under general. It's the most impactful feature (and takes up the most space in the panel) so it deserves to be right up there. 

https://github.com/invoke-ai/InvokeAI/assets/4822129/594bf724-0110-4734-9e03-2471a2ca4698

**Pull request merged**